### PR TITLE
[#137] Proposal

### PR DESCRIPTION
**Pull request by pasqualeorsini**

### CODE REVIEW

**Comment by fbiville** [comment]
@pasqualeorsini any help needed? Let me know!

**Comment by pasqualeorsini** [comment]
@fbiville It's starting to take shape. I'm hoping to have something to show by the end of the week.

**Comment by pasqualeorsini** [comment]
@fbiville It's done for now.  Can you have a look please?

**Comment by fbiville** [comment]
Awesome!! Let me have a look :)

**Comment by fbiville** [comment]
I think you'll have to rebase your changes against `master_3.x` instead of `master`.

**Comment by fbiville** [comment]
@pasqualeorsini I can offer either of the following:

- send a PR to your fork with suggestions
- provide detailed feedback on your current PR

which one do you prefer? if you don't have much time, I'm totally fine with doing the first.

**Comment by pasqualeorsini** [comment]
Giving a detailed feedback on this PR is fine. I'm all ears. :-)

File: examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/Application.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    public static void main(String... args) throws Exception {
+        MigrationComponent migrationComponent = DaggerMigrationComponent.create();
+        Application application = new Application();
+        application.executeSimpleMigration(migrationComponent);
+    }
+
+    public void executeSimpleMigration(MigrationComponent migrationComponent) {
+        // Running the migration
+        migrationComponent.liquigraph();
```
Comment (fbiville) [inline examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/Application.java]: as you'll see below in the configuration class, I would rather transform this into:

```java
Liquigraph liquigraph = migrationComponent.liquigraph();
liquigraph.runMigrations(migrationComponent.configuration());
```

File: examples/dagger2/src/test/java/org/liquigraph/examples/dagger2/MigrationWebIT.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.junit.Test;
+import org.liquigraph.examples.dagger2.di.component.DaggerMigrationTestComponent;
+
+import static io.restassured.RestAssured.get;
+import static org.hamcrest.Matchers.containsString;
+
+public class MigrationWebIT {
+    @Test
+    public void testMigration() throws Exception {
```
Comment (fbiville) [inline examples/dagger2/src/test/java/org/liquigraph/examples/dagger2/MigrationWebIT.java]: please rename to `verifies_service_responds_after_migration`

**Comment by fbiville** [comment]
@pasqualeorsini I can help with the changes if you need ;-)

**Comment by pasqualeorsini** [comment]
@fbiville Thanks for the feedback, it's always enlightening to have another pair of eyes. :-) 
I'm going to make the changes you proposed and if I need help I'll let you know for sure.

**Comment by fbiville** [comment]
Hi @pasqualeorsini I would like the examples to be integrated before I release. Do you know when you could work on this?

**Comment by pasqualeorsini** [comment]
@fbiville The work is mostly done. 
I have one last concern regarding the in-process Neo4J instance : I didn't manage to find out how to set the port.

Do I need to open another pull request to rebase on master_3.x from my fork's master_3.x?

**Comment by fbiville** [comment]
@pasqualeorsini you can rebase your current fork's branch and `git push --force-with-lease`, no need to open another PR

File: examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/Application.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        MigrationComponent migrationComponent = DaggerMigrationComponent.create();
+        Application application = new Application();
+        application.executeSimpleMigration(migrationComponent);
+    }
+
+    public void executeSimpleMigration(MigrationComponent migrationComponent) {
+
+        Liquigraph liquigraph = migrationComponent.liquigraph();
+
+        // Running the migrations
```
Comment (fbiville) [inline examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/Application.java]: just a minor detail but the code below the comment is quite self-explanatory, I think, so the comment could be removed

File: examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/Application.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        liquigraph.runMigrations(migrationComponent.configuration());
+
+        // Starting the web server
```
Comment (fbiville) [inline examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/Application.java]: same here

File: examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/dao/SentenceRepository.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package org.liquigraph.examples.dagger2.dao;
+
+import org.liquigraph.examples.dagger2.entity.Sentence;
+
+import java.util.Optional;
+
+public interface SentenceRepository {
+
+    Optional<Sentence> findOne(String cypherQuery);
```
Comment (fbiville) [inline examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/dao/SentenceRepository.java]: actually: my previous signature proposal is a bit weird (sorry :sweat_smile:), I think
```java
Optional<Sentence> findOne()
```
would make more sense (no need to expose the query to the client)

File: examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/di/component/MigrationComponent.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.inject.Singleton;
+
+/**
+ * Liquigraph and SentenceRepository using a real neo4j database
+ */
+@Singleton
+@Component(modules = {LiquigraphModule.class})
+public interface MigrationComponent {
+    SentenceRepository dao();
```
Comment (fbiville) [inline examples/dagger2/src/main/java/org/liquigraph/examples/dagger2/di/component/MigrationComponent.java]: minor detail, but I think the method should be named `sentenceRepository` instead

File: examples/dagger2/src/test/java/org/liquigraph/examples/dagger2/MigrationWebIT.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+;
+
+public class MigrationWebIT {
+    @Test
+    public void verifies_service_responds_after_migration() throws Exception {
+        ServerControls serverControls = null;
+        try {
+            TestServerBuilder testServerBuilder = TestServerBuilders.newInProcessBuilder();
+            System.setProperty("java.io.tmpdir", "target/neo4j-hello-db");
+            testServerBuilder.newServer();
```
Comment (fbiville) [inline examples/dagger2/src/test/java/org/liquigraph/examples/dagger2/MigrationWebIT.java]: you can move this to a `@Before`-annotated method

File: examples/dagger2/src/test/java/org/liquigraph/examples/dagger2/MigrationWebIT.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+            application.executeSimpleMigration(migrationComponent);
+
+            get("/")
+                    .then()
+                    .assertThat()
+                    .body(containsString("Hello world!"));
+        } finally {
+            if (serverControls != null)
+                serverControls.close();
```
Comment (fbiville) [inline examples/dagger2/src/test/java/org/liquigraph/examples/dagger2/MigrationWebIT.java]: and that to an `@After` method

**Comment by pasqualeorsini** [comment]
@fbiville I opened a new PR to rebase from the master_3.x branch : #160

**Comment by fbiville** [comment]
Superseded by https://github.com/fbiville/liquigraph/pull/160

**Pull request closed without merging**